### PR TITLE
Feature.packaging enhancement 980

### LINKED ISSUE #980: RPM/DEB build automation
This issue is to address the creation of automation around building .rpm and .deb packages natively within the different operating systems that are supported.  This automation is to auto-handle the:

- The resolution of dependencies
  - stdeb.cfg (.deb building)
  - setup.cfg (.rpm building)
- Building of the packages
- The Testing of the packages

As such, this new solution should include a means by which to make a 'one source of truth' source file for all dependencies for a project.

Since this automation is a heavy lift, this automation should also be...

- Transferable to other projects and other github repos
- Modifiable and future proof

### PR DESCRIPTION
**Pull request by ssorenso**
Issues:
Fixes #980 
Via peripheral, this may be (and can be edited later if it does not) fixing:
#915 
#849 
#596 

Problem:
The issues that this addresses are as follows:

- Introduces a 'single source of truth' file location for dependencies in a supportable manner
  - The `setup.py` introduces issues with attempting to import and extract values without the Setup Utils taking over and attempting to apply a build
  - Issue addressed here is #596 
  - This can be migrated later to a different solution as needed
    - I've isolate this code to the stand-alone script `construct_setups.py`
      - This addresses both the construction of the `setup.cfg` and the `stdeb.cfg` files (rpm and deb respectuflly)
      - Other solutions that might work would be a `pipfile`, `tox.ini`, etc.
      - I used the requirements.txt format because it was the easiest to parse by hand, which is a requirement on the testing side
      - Making testing as bare-to-the-native OS as possible for testing purposes and best validation for customers who might not have the parsing tools installed
- Introduces a more generic means of testing packages that is not environment-specific
  - Thus anyone could put their git repo clone anywhere on their machine, and running this will simply build and test
  - This also allows you to take the code and port it over into a different github repo and run it, and it should work there as well
    - As long as the `*-dist/scripts` and `*-dist/Docker/<os>/<os_version>` and `*-dist/Docker/<os>/install_test/` structures are in place
- Performs auto-handling of 'ceiling' requirements for dependency versions
- Integrates well with `.travis.yml` testing as it should exit with a non-zero status within the installation portion of the Travis processing
  - Thus, it should notify a pull requestor if/when the package building portion breaks

I'm not saying that this solves all problems here for all folks involved's requirements; however, we can say that this is a good starting point.  That is to handle building and testing the packages.

Analysis:
The following address the following issues and have the following specifications:
- The `.travis.yml` change
  - Puts the one build/test script at the `installation` level
    - Thus any non-0 exiting script should cause a 'break' in the test
- The `<project>-dist/<os>/<version>/*` changes
  - Introduces a more generic means of building the packages in their native operating system
  - Has a higher level of sustainability as they are programmed in python, and the Docker file's generated containers should exit with a non-zero when things fail
  - Is executed from the build script via `subprocess.check_call()` and handled to report any failures via a non-0 exit status
- The `<project>-dist/<os>/install_test/*` changes
  - Introduces a more sustainable version of the scripts that are not as argument dependent
  - Introduces a more generic handling of dependencies such that a package might require any number of tiers of packages from both F5 and the Linux community (`apt-get` or `yum`)
- The `<project>-dist/scripts/` introduces a go-to repository of build script material for importing and handling
  - The `build_exceptions.py`
    - A generic set of exceptions that use `inspect` traceback and frame handling to report on errors that are predictable
    - A means to grow this list of exceptions as needed via inheritence
    - At each level, uniquely-identifiable changes can be performed on the exceptions to make things more usable to programmers
  - The `build_pkgs.py`
    - A more generic build and test script that can be deployed (if with all of its requirements) to other repos and github repositories as needed
  - The `constructs_setup.py`
    - A standalone script that can be called on its own or used as a modulino to be imported by another script (how it's used by `build_pkgs.py`)
    - This uses the single source of truth dependencies file to generate a `setup.cfg` file for RPM's and a `stdeb.cfg` file for DEB's
    - Allows for version ranges for dependencies (both a ceiling and a floor) as needed.
  - The `configure.py`
    - Constructs a `<project>-dist/scripts/config.JSON` file that contains globals
    - This is the lever that is used throughout the rest of the code to identify where in file space the runtime is
  - The `add_pkg_name.py`
    - Script that allows the build scripts to "talk" to the test scripts communicating what the package names are
    - Depends upon the `configure.py` as a modulino that houses all things `config.JSON`
    - Can be made to lock the `config.JSON` file in the future for parallel processing the RPM and DEB builds/tests
  - The `build_expectations.py`
    - A simple class-holding module that contains `Build()` object
    - This object is used to derive a tuple([namedtuple('OperationSettings')]) as well as the `namedtuple(OperationSettings)` objects
   - These objects are used as a means of uniforming the format of what is used to determine what will be built/tested by both the `build_pkgs.py` and the `isntall_test.py`
  - The `install_test.py`
    - This too can be standalone to test the already-present packages
    - This script is used as a module by the `build_pkgs.py` script
    - Executes the tests to test the packages in their native OS/OS version 
- The `setup_requirements.txt` file as a "scarecrow" source of truth file for dependencies
  - Easily parsable via just `re` and also has a `pip.req.parse_requirements` function to parse this data as well

Tests:
This code, in and of itself, is a test; however, Travis is what executes it.

SIDE NOTE:
This code is broken out into several commits intentionally to address to the reviewer the code and what it does in smaller chunks.

Reviewer: @zancas

### CODE REVIEW

File: f5-sdk-dist/README.rst [hunk 1]
```diff
@@ -0,0 +1,77 @@
+f5 Package Creation and Automation
+==================================
+
+Introduction
+------------
+You've gon the wrong way!  This directory and lower directories are not
```
Comment (jlongstaf) [inline f5-sdk-dist/README.rst:6]: Typo: 'gon' missing 'e'.

File: f5-sdk-dist/scripts/README.rst [hunk 2]
```diff
@@ -0,0 +1,77 @@
+f5 Package Creation and Automation
+==================================
+
+Introduction
+------------
+You've gon the wrong way!  This directory and lower directories are not
```
Comment (caphrim007) [inline f5-sdk-dist/scripts/README.rst:6]: type here too

**Comment by ssorenso** [comment]
So this is runnable from a machine that has Docker installed, and the expected CLI feedback is:
```
f5-common-python$ ./f5-sdk-dist/scripts/build_pkgs.py
Successfully constructed /git/f5-common-python/setup.cfg and
/git/f5-common-python/f5-sdk-dist/deb_dist/stdeb.cfg
Building packages...
For Debian...
For Redhat...
Completed package builds...
Initiating install tests...
Testing Packages...
Package: ./f5-sdk-dist/deb_dist/python-f5-sdk_2.2.2-1_1404_all.deb Test succeeded!
Package: ./f5-sdk-dist/rpms/build/f5-sdk-2.2.2-1.el7.noarch.rpm Test succeeded!
Completed install tests
```

If something "bad" happens, it will be much more verbose if it fails during the testing or building phases for a package.  Further, it should always produce a Traceback upon failure.

Running this via time to get a clock count, I've got:
```
f5-common-python$ time !!
time ./f5-sdk-dist/scripts/build_pkgs.py
Successfully constructed /git/f5-common-python/setup.cfg and
/git/f5-common-python/f5-sdk-dist/deb_dist/stdeb.cfg
Building packages...
For Debian...
For Redhat...
Completed package builds...
Initiating install tests...
Testing Packages...
Package: ./f5-sdk-dist/deb_dist/python-f5-sdk_2.2.2-1_1404_all.deb Test succeeded!
Package: ./f5-sdk-dist/rpms/build/f5-sdk-2.2.2-1.el7.noarch.rpm Test succeeded!
Completed install tests

real	0m49.109s
user	0m0.381s
sys	0m0.279s
```

So just under a minute.  However, it will likely take longer for a machine that does not have a pre-existing docker image set for this... It takes roughly 2-3 minutes in the Travis testing to do so.  Let me know if you want me to test this from scratch.

File: f5-sdk-dist/README.rst (hunk 1 above)
Comment (ssorenso) [reply f5-sdk-dist/README.rst:6]: Cool!  Nice catch!  I've got a fix ready to push for this

File: f5-sdk-dist/scripts/README.rst (hunk 2 above)
Comment (ssorenso) [reply f5-sdk-dist/scripts/README.rst:6]: Cool!  Nice catch!  I've got a fix ready to push for this

File: .travis.yml [hunk 3]
```diff
@@ -25,15 +25,12 @@ script:
   - tox -e flake
   - tox -e unit
   - python setup.py sdist
-  - ${DIST_REPO}/scripts/package.sh "redhat" "7"
-  - ${DIST_REPO}/scripts/package.sh "ubuntu" "14.04"
+  - ${DIST_REPO}/scripts/build_pkgs.py  # ubuntu;14.04,15.01 # basically <os>;<version> !! overwrites !!
```
Comment (zancas) [inline .travis.yml:28]: Should this comment mention redhat?

File: f5-sdk-dist/Docker/ubuntu/14.04/Dockerfile [hunk 4]
```diff
@@ -4,7 +4,12 @@ FROM ubuntu:trusty
...
 	fakeroot \
-	python-all
-
-COPY ./build-debs.sh /
+	python-all \
+    python-pip
 
+COPY ./build-debs.py /build-debs.py
+RUN chmod 777 /build-debs.py
+RUN ls /build-debs.py
```
Comment (zancas) [inline f5-sdk-dist/Docker/ubuntu/14.04/Dockerfile:12]: Is this from debugging the Dockerfile?

File: f5-sdk-dist/Docker/ubuntu/14.04/build-debs.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env python
```
Comment (zancas) [inline f5-sdk-dist/Docker/ubuntu/14.04/build-debs.py:1]: I vote we name this file "`build-deb.py`"

File: f5-sdk-dist/Docker/ubuntu/14.04/build-debs.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            config = os.path.dirname(item) + "/config.JSON"
+            if os.path.isfile(config):
+                configure = item
+                break
+    if os.path.isfile(config) and os.path.isfile(configure):
+        subprocess.check_output(str("python %s" % configure).split())
+        with open(config, 'r') as fh:
+            config = json.loads(fh.read())
+    else:
+        raise EnvironmentError("Exp setup parameters missing!")
```
Comment (zancas) [inline f5-sdk-dist/Docker/ubuntu/14.04/build-debs.py:48]: "Exp"  ---> "Expected"

Comment (zancas) [reply f5-sdk-dist/Docker/ubuntu/14.04/build-debs.py:48]: Maybe say which one's missing in the exception.

File: f5-sdk-dist/Docker/ubuntu/14.04/build-debs.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    version = \
+        subprocess.check_output([cmd], shell=True).strip()
+    tmp_dist = "/var/deb_dist"
+    project = config['project']
+    tmp_dist = "/var/deb_dist"
+    os_version = "1404"
+    deb_dir = "%s/deb_dist" % config['dist_dir']
+    print("Building %s debian packages..." % project)
+    shutil.copyfile("%s/stdeb.cfg" % (deb_dir), "./stdeb.cfg")
+    shutil.copytree(deb_dir, tmp_dist)
```
Comment (zancas) [inline f5-sdk-dist/Docker/ubuntu/14.04/build-debs.py:71]: Add a comment saying end of config block, or assignment?

**Review by jlongstaf — APPROVED** [review]
Nice work. Nothing to suggest other than fixing typos.

File: f5-sdk-dist/scripts/configure.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+#
+
+import errno
+import glob
+import json
+import os
+import re
+import sys
+
+
```
Comment (zancas) [inline f5-sdk-dist/scripts/configure.py:24]: How about something like:

```
DISTRIBUTION_ROOT = <your method for finding and creating the abspath for the directory with setup.py in it.

def _retrieve_and_validate_dist_dir(dist_root):
    dist_dirs = [x for x in os.listdir(dist_root) if (os.path.isdir(x) and x.endswith('dist'))]
    if len(dist_dirs) != 1:
        raise AppropriateException
    return dist_dirs[0]

DISTRIBUTION_DIRECTORY = _retrieve_and_validate_dist_dir(DISTRIBUTION_ROOT)

```
Then import DISTRIBUTION_DIRECTORY as necessary?

File: f5-sdk-dist/scripts/configure.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    """
+    kvp = {pkg_type: pkg_name}
+    working = os.getcwd() if not working else os.path.abspath(working)
+    if '-dist/script' in working:
+        config = working + "/config.JSON"
+    elif '-dist' in working:
+        config = working + "/script/config.JSON"
+    else:
+        config = working + "/*-dist/script/config.JSON"
+        entropy = glob.glob(config)
```
Comment (zancas) [inline f5-sdk-dist/scripts/configure.py:40]: The use of `entropy` is thought provoking and I like that.   I think the cost to readability should be considered.   Should every reader of this code be required to understand the concept of entropy?   What's the purpose of the variable name?

File: f5-sdk-dist/scripts/README.rst (hunk 2 above)
Comment (zancas) [reply f5-sdk-dist/scripts/README.rst:6]: The "You've gon the wrong way!" warning is cute, but maybe: "Have you gone the wrong way?!"

**Review by zancas — APPROVED** [review]
LGTM.

**Pull request merged**